lightning/src/routing/router.rs

```diff
@@ -1498,6 +1501,55 @@ where L::Target: Logger {
 		}
 	}
 
+	// Step (10).
+	// Add a random 'shadow route' offset to the CLTV expiry delta values to improve privacy.
+	//
+	// Start from the payee
+	let mut cur_node_id = payee_node_id;
+	let mut shadow_ctlv_expiry_delta_offset: u32 = 0;
+
+	let mut prng = ChaCha20::new(random_bytes, &[0; 8]);
+	let mut random_index_bytes = [0u8; 8];
+	let mut random_index: usize;
+
+	// Do a three-hop random walk
+	for _random_hop in 0..3 { 
+		if let Some(cur_node) = network_nodes.get(&cur_node_id) {
```

Comment:
This is great, but I think we should handle the case where the last hop is unknown to us, at least. Consider when you're paying a mobile node through a large LSP, the last-hop will ~never be in the network grpah, but we do want to fudge things starting from the second-to-last hop, pretending we're routing through them. To handle that we'll need to tweak quite a bit here, we may need to calculate such random walks per-path, not just in total. That said, I think we need to *avoid* doing things per-path if the last hop is public - its totally possible the second-to-last public hop will be the same across two paths, which will cause us to tell that node explicitly that we're faking the CLTVs, given they're different on the same hash HTLC.

Reply:
Ah, right, I didn't account for MPP and private channels here. Maybe a good way to fix this would be to always apply the offset on a per-path basis, but set the starting point to the predecessor of `payee` when itself is not present in the public graph. The randomness used for the random walk could then come from something like `random_bytes XOR cur_node_id`, which would make the offset the same if all paths share the same starting point and different on a per-path basis otherwise.

Reply:
Yea, that seems clever, I like that idea.

Reply:
I now simply used the first 8 bytes of `cur_node_id` as the `ChaCha20` nonce, which should give us the wanted semi-determinism. See https://github.com/lightningdevkit/rust-lightning/commit/ea0ad9d079ba51e5cf6f780999661a1b7673351b.